src/storage/index/IndexExecutor.inl

```diff
@@ -0,0 +1,329 @@
+                                                   spaceId_,
+                                                   tagOrEdge_);
+        auto row = getRowFromReader(reader.get());
+        data->set_props(std::move(row));
+    } else {
+        VLOG(3) << "Missed partId " << partId
+                << ", src " << src << ", edgeType "
+                << tagOrEdge_ << ", Rank "
+                << rank << ", dst " << dst;
+        return kvstore::ResultCode::ERR_KEY_NOT_FOUND;
+    }
+    return ret;
+}
+
+template<typename RESP>
+std::string IndexExecutor<RESP>::getRowFromReader(RowReader* reader) {
+    RowWriter writer;
+    PropsCollector collector(&writer);
+    this->collectProps(reader, props_, &collector);
+    return writer.encode();
+}
+
+template<typename RESP>
+bool IndexExecutor<RESP>::conditionsCheck(const folly::StringPiece& key) {
+    UNUSED(key);
+    Getters getters;
+    getters.getAliasProp = [this, &key](const std::string&,
+                                        const std::string &prop) -> OptVariantType {
+            return decodeValue(key, prop);
+    };
+    return exprEval(getters);
+}
+
+template<typename RESP>
+OptVariantType IndexExecutor<RESP>::decodeValue(const folly::StringPiece& key,
+                                                const folly::StringPiece& prop) {
+    using nebula::cpp2::SupportedType;
+    auto type = indexCols_[prop.str()];
+    return NebulaKeyUtils::decodeVariant(getIndexVal(key, prop), type);
+}
+
+template<typename RESP>
+folly::StringPiece IndexExecutor<RESP>::getIndexVal(const folly::StringPiece& key,
+                                                    const folly::StringPiece& prop) {
+    auto tailLen = (!isEdgeIndex_) ? sizeof(VertexID) :
+                                     sizeof(VertexID) * 2 + sizeof(EdgeRanking);
+    using nebula::cpp2::SupportedType;
+    size_t offset = sizeof(PartitionID) + sizeof(IndexID);
+    size_t len = 0;
+    int32_t vCount = vColNum_;
+    for (const auto& col : index_->get_fields()) {
+        switch (col.get_type().get_type()) {
+            case SupportedType::BOOL: {
+                len = sizeof(bool);
+                break;
+            }
+            case SupportedType::TIMESTAMP:
+            case SupportedType::INT: {
+                len = sizeof(int64_t);
+                break;
+            }
+            case SupportedType::FLOAT:
+            case SupportedType::DOUBLE: {
+                len = sizeof(double);
+                break;
+            }
+            case SupportedType::STRING: {
+                auto off = key.size() - vCount * sizeof(int32_t) - tailLen;
+                len = *reinterpret_cast<const int32_t*>(key.data() + off);
+                --vCount;
+                break;
+            }
+            default:
+                len = 0;
+        }
+        if (col.get_name() == prop.str()) {
+            break;
+        }
+        offset += len;
+    }
+    return key.subpiece(offset, len);
```

Comment:
Need to check the` key.len` with `offset + len`, if not equal, need return error

Reply:
These are never going to be equal.